### PR TITLE
Allow init scripts to apply plugins to settings

### PR DESCRIPTION
**Pull request by ldaley**
Upcoming work will move a key plugin of ours to being settings based instead of project based. This PR makes it practical to continue to use an init script to apply the plugin across “projects” (i.e. in the logical sense), by opening up a callback for accessing the settings object before the settings script is applied.

Documentation, other than API doc, has not been added as this is a rather niche feature. The documentation for the plugin itself will provide a recipe for such usage.

https://builds.gradle.org/project.html?projectId=Gradle&tab=projectOverview&branch_Gradle=ldaley%2Fbefore-settings-lifecycle-callback

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/invocation/Gradle.java [hunk 1]
```diff
@@ -190,6 +190,24 @@
...
     void buildStarted(Action<? super Gradle> action);
 
+    /**
+     * Adds an action to be called before the build settings have been loaded and evaluated.
+     *
+     * @param closure The action to execute.
+     * @since 5.7
+     */
+    @Incubating
+    void beforeSettings(Closure<?> closure);
```
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/api/invocation/Gradle.java]: Isn't the `Action`-taking variant enough?

File: subprojects/core-api/src/main/java/org/gradle/BuildAdapter.java [hunk 2]
```diff
@@ -27,6 +27,10 @@
     public void buildStarted(Gradle gradle) {
     }
 
+    @Override
+    public void beforeSettings(Settings settings) {
+    }
+
```
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/BuildAdapter.java]: Is this required given that the interface has a default implementation for this method?

File: subprojects/core-api/src/main/java/org/gradle/api/invocation/Gradle.java (hunk 1 above)
Comment (ldaley) [reply subprojects/core-api/src/main/java/org/gradle/api/invocation/Gradle.java]: Unfortunately not. Please see: https://gradle.slack.com/archives/C03A0MT6S/p1566182271414700

TL;DR: these closure methods behave different to most.

File: subprojects/core-api/src/main/java/org/gradle/BuildAdapter.java (hunk 2 above)
Comment (ldaley) [reply subprojects/core-api/src/main/java/org/gradle/BuildAdapter.java]: It's not, but I thought it best to add it for consistency.

**Comment by ldaley** [comment]
@big-guy ping.

**Comment by ldaley** [comment]
@big-guy will wait for @lptr to 👍 , then bump to 6.0 and merge.

**Comment by lptr** [comment]
LGTM!
